src/components/application_manager/include/application_manager/state_controller_impl.h

```diff
@@ -152,12 +152,17 @@ class StateControllerImpl : public event_engine::EventObserver,
 
   template <typename UnaryFunction>
   void ForEachApplication(UnaryFunction func) const {
-    DataAccessor<ApplicationSet> accessor = app_mngr_.applications();
-    ApplicationSet::iterator it = accessor.GetData().begin();
-    for (; it != accessor.GetData().end(); ++it) {
-      ApplicationConstSharedPtr const_app = *it;
-      if (const_app) {
-        func(app_mngr_.application(const_app->app_id()));
+    ApplicationSet applications;
+    {
+      DataAccessor<ApplicationSet> accessor = app_mngr_.applications();
+      applications = accessor.GetData();
+    }
+
+    ApplicationSet::iterator it = applications.begin();
+    for (; it != applications.end(); ++it) {
+      auto app = *it;
+      if (app) {
```

Comment:
@IGapchuk you don't need this check anymore. Because you copied applications collection of shared ptrs, `ApplicationImpl` instance will not be destroyed until `applications` variable is destroyed as it also holds the same shared ptrs. That's why all pointers in this loop will be valid. You can remove this check

Reply:
@AKalinich-Luxoft Removed in c874387b10610d059a863f43497779f73aaa26e7